### DISCUSSION TITLE
Proposal: NotNullWhenArgumentIsNonZero attribute

### DISCUSSION THREAD

**Discussion by NN---** [dotnet/csharplang#3517] [Language Ideas]
## Background and Motivation

```cs
public static string? PtrToStringAuto(IntPtr ptr)
        {
            return PtrToStringUni(ptr);
        }
```

PtrToStructure methods should be annotated as returning non null when pointer is not null

```
A managed string that holds a copy of the unmanaged string if the value of the ptr parameter is not null; otherwise, this method returns null.
```
https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.marshal.ptrtostringauto?view=netcore-3.1

Proposal:
Introduce some attribute like:

```cs
[NotNullWhenArgumentIsNonZero]
public static string? PtrToStringAuto(IntPtr ptr)
        {
            return PtrToStringUni(ptr);
        }
```

**Comment by NN---** [comment]
Perhaps general `NotNullWhenNotNull(“ptr”)]` is even better

**Comment by 333fred** [comment]
We talked about additional overloads to NotNullWhen/MaybeNullWhen for error codes during initial release, but didn't see high enough usage to warrant. However, we did want to see if the community wanted them. @jcouv, I believe you've been tracking these requests?

> Perhaps general NotNullWhenNotNull(“ptr”)] is even better 

What's the difference between that and NotNullIfNotNull?

**Comment by NN---** [comment]
Ah, right there is such one and on the other hand it doesn’t cover this case.

For IntPtr? there are 3 value options , not 2.

**Comment by alrz** [comment]
>What's the difference between that and NotNullIfNotNull?

I think that's to say `NotNullIfNotDefault` because IntPtr is a struct.

**Comment by 333fred** [comment]
> I think that's to say `NotNullIfNotDefault` because IntPtr is a struct.

Ah, well that's really another big feature request the current flow analysis isn't set up to track. We track null vs not null, not whether a value type is default or not.

**Comment by alrz** [comment]
FYI there's a proposal for that: #146 (looks like it's been rejected)

**Comment by Joe4evr** [comment]
@alrz It's not quite rejected as it is considered to be done as part of #99.
